x/ibc/04-channel/client/utils/utils.go

```
@@ -101,67 +103,93 @@ func queryChannelABCI(clientCtx client.Context, portID, channelID string) (*type
+			return nil, err
+		}
+
+		// use client state returned from ABCI query in case query height differs
+		identifiedClientState := clienttypes.NewIdentifiedClientState(res.IdentifiedClientState.ID, clientState)
+		res = types.NewQueryChannelClientStateResponse(identifiedClientState, proof, int64(proofHeight))
 	}
-	return clientState, height, nil
+
+	return res, nil
 }
 
-// QueryChannelConsensusState uses the channel Querier to return the ConsensusState
-// of a Channel.
-func QueryChannelConsensusState(clientCtx client.Context, portID, channelID string) (clientexported.ConsensusState, int64, error) {
-	params := types.NewQueryChannelConsensusStateRequest(portID, channelID)
-	bz, err := clientCtx.JSONMarshaler.MarshalJSON(params)
-	if err != nil {
-		return nil, 0, fmt.Errorf("failed to marshal query params: %w", err)
+// QueryChannelConsensusState returns the ConsensusState of a channel end. If
+// prove is true, it performs an ABCI store query in order to retrieve the
+// merkle proof. Otherwise, it uses the gRPC query client.
+func QueryChannelConsensusState(
+	clientCtx client.Context, portID, channelID string, prove bool,
+) (*types.QueryChannelConsensusStateResponse, error) {
+
+	queryClient := types.NewQueryClient(clientCtx)
+	req := &types.QueryChannelConsensusStateRequest{
+		PortID:    portID,
+		ChannelID: channelID,
 	}
 
-	route := fmt.Sprintf("custom/%s/%s", types.QuerierRoute, types.QueryChannelConsensusState)
-	res, height, err := clientCtx.QueryWithData(route, bz)
+	res, err := queryClient.ChannelConsensusState(context.Background(), req)
 	if err != nil {
-		return nil, 0, err
+		return nil, err
 	}
 
-	var consensusState clientexported.ConsensusState
-	err = clientCtx.JSONMarshaler.UnmarshalJSON(res, &consensusState)
-	if err != nil {
-		return nil, 0, fmt.Errorf("failed to unmarshal consensus state: %w", err)
+	consensusState, ok := res.ConsensusState.GetCachedValue().(clientexported.ConsensusState)
+	if !ok {
+		panic("invalid consensus state")
```

I think this should return an error instead?